parachain/pallets/outbound-queue/src/lib.rs

```diff
@@ -314,94 +397,46 @@ pub mod pallet {
 			T::GasMeter::MAXIMUM_BASE_GAS + T::GasMeter::maximum_required(command)
 		}
 
-		/// Fee in DOT for delivering a message.
-		pub(crate) fn delivery_fee(command: &Command) -> T::Balance {
-			let max_gas_used = Self::maximum_overall_required_gas(command);
-			T::DeliveryFeePerGas::get().saturating_mul(max_gas_used.into())
-		}
-
-		/// Maximum refund in Ether for delivering a message
-		pub(crate) fn maximum_refund(command: &Command) -> u128 {
-			let max_gas_used = Self::maximum_overall_required_gas(command);
-			T::DeliveryRefundPerGas::get().saturating_mul(max_gas_used as u128)
-		}
-	}
-
-	impl<T: Config> OutboundQueueTrait for Pallet<T> {
-		type Ticket = OutboundQueueTicket<MaxEnqueuedMessageSizeOf<T>>;
-		type Balance = T::Balance;
-
-		fn validate(message: &Message) -> Result<(Self::Ticket, Fees<Self::Balance>), SendError> {
-			// The inner payload should not be too large
-			let payload = message.command.abi_encode();
-
-			// Create a message id for tracking progress in submission pipeline
-			let message_id: MessageHash = sp_io::hashing::blake2_256(&(message.encode())).into();
-
-			ensure!(
-				payload.len() < T::MaxMessagePayloadSize::get() as usize,
-				SendError::MessageTooLarge
+		/// Calculate fee in native currency for delivering a message.
+		pub(crate) fn calculate_fee(command: &Command) -> Option<Fee<T::Balance>> {
+			let max_gas = Self::maximum_overall_required_gas(command);
+			let remote_fee = Self::calculate_remote_fee(
+				max_gas,
+				Self::fee_config().fee_per_gas,
+				Self::fee_config().reward,
 			);
-
-			let base_fee = T::WeightToFee::weight_to_fee(
-				&T::WeightInfo::do_process_message()
-					.saturating_add(T::WeightInfo::commit_one_message()),
-			);
-			let delivery_fee = Self::delivery_fee(&message.command);
-
-			let command = message.command.clone();
-			let fee = Fees { base: base_fee, delivery: delivery_fee };
-
-			let enqueued_message: EnqueuedMessage =
-				EnqueuedMessage { id: message_id, origin: message.origin, command };
-			// The whole message should not be too large
-			let encoded =
-				enqueued_message.encode().try_into().map_err(|_| SendError::MessageTooLarge)?;
-
-			let ticket =
-				OutboundQueueTicket { id: message_id, origin: message.origin, message: encoded };
-
-			Ok((ticket, fee))
+			let remote_fee = FixedU128::from(remote_fee)
+				.checked_div(&Self::fee_config().exchange_rate)?
+				.into_inner()
+				.checked_div(FixedU128::accuracy())
+				.expect("accuracy is not zero; qed")
+				.into();
```

Comment:
If we add `exchange_rate` into the calculation, we also need to take token precision into consideration, DOT is 10 while Ether 18.

Reply:
Nice catch! This means we should divide the result again by `1 * 10*^-(18 - 10)` right?

Reply:
Fixed in https://github.com/Snowfork/snowbridge/pull/982/commits/a0d7916624012f2b32c4c6ce846e8ba092c6d5bb